lib/Sema/CSClosure.cpp

```diff
@@ -885,11 +885,13 @@ class SyntacticElementConstraintGenerator
     } else {
       // If this is simplify `return`, let's create an empty tuple
       // which is also useful if contextual turns out to be e.g. `Void?`.
-      resultExpr = getVoidExpr(cs.getASTContext());
+      // Also, attach return stmt source location so if there is a contextual
+      // mismatch we can produce a diagnostic in a valid source location.
+      resultExpr = getVoidExpr(cs.getASTContext(), returnStmt->getEndLoc());
     }
 
     SolutionApplicationTarget target(resultExpr, context.getAsDeclContext(),
-                                     CTP_ReturnStmt, resultType,
+                                     CTP_ClosureResult, resultType,
```

Comment:
@LucianoPAlmeida I have been testing my AST transform for result builders and apparently this change actually breaks opaque result type opening - ConstraintSystem::openOpaqueType only accepts `CTP_ReturnStmt` for opening. I think we should change locators back and modify diagnostics instead to detect that CTP_ReturnStmt comes from the closure body. Would you like to work on that or should I?

Reply:
Ah sorry for that. I thought CPT use was most for diagnostics, I can work on it. First I'll revert this just to umbrela things and try to adjust that separately :)

Reply:
Do you have a repro? I thought it would be something like this:
```swift
func f() -> some BinaryInteger {
  return ["foo"].map { s in
       if s == "1" { return } // expected-error{{cannot convert value of type '()' to closure result type 'Bool'}}
       return s.count
  }
}
```
Just want to add a regression test

Reply:
So I tried to work it out and unfortunately it doesn't seem possible without my changes because `SyntacticElementConstraintGenerator` only handles closures at the moment but with my changes it would be able to handle accessors (mutate their bodies to apply result builder and what's where that issue happens). Note: it's okay to use `CTP_ClosureResult` if "context" is a closure but `CTP_ReturnStmt` should be used in all other cases.

You can build https://github.com/apple/swift/pull/60065 locally and test the following snippet with the feature enabled (`-enable-experimental-feature ResultBuilderASTTransform`) to reproduce:

```
import SwiftUI

struct EqualWitdthHStack : Layout {
  func sizeThatFits(proposal: ProposedViewSize, subviews: Self.Subviews, cache: inout Self.Cache) -> CGSize { fatalError() }
  func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Self.Subviews, cache: inout ()) {}
}

struct MyView : View {
  var body : some View {
    EqualWitdthHStack {
      EmptyView()
    }
  }
}
```

Type-checking of that with `ReturnStmt` target set to use `CTP_ClosureResult` would result in `error: cannot convert value of type 'some View' (result of 'Self.callAsFunction') to closure result type 'some View' (type of 'MyView.body')`

Reply:
@xedin Quick question about the approach you mentioned here

>  I think we should change locators back and modify diagnostics instead to detect that CTP_ReturnStmt comes from the closure body

What do you mean by changing locators back? Track this information via locator?

Reply:
Oh I forgot about this one, this week has been busy. But I'll probably wait for that PR to land before work on this

Reply:
Hey @xedin, trying to introduce this test snippet but it looks like failing with `cannot find type 'Layout' in scope`. I don't know much about SwiftUI so is that a type a `SwiftUI` type added in a specific version? Do I have to have a specific version of OS with that type to run that locally? 
Just not sure about what am I missing 😅

Reply:
Ah actually never mind, I just found that some of the types in the snippet are really `macOS 13.0+` so now everything makes sense again =]
Do you have a repro snippet that doesn't need those?